_posts/update-me-please.md

```diff
@@ -0,0 +1,157 @@
+
+## IT/Ops facilities 
+
+A notification service is not a genuine business service in SOA terminology. It's an IT/Ops kind of service that provides facilities to business services in the system. In the case of this article, the notification service offers business services a way to deliver notifications. To do so, it ships a notifications package that business services can use to interact with it. The notifications package is composed of an `INotificationService` interface that looks like the following:
+
+```csharp
+interface INotificationService
+{
+   Task<IEnumerable<Subscription>> GetSubscriptionsFor(string eventId);
+   Task Dispatch(IEnumerable<Notification> notifications);
+}
+```
+
+And the `Subscription` and `Notification` types could look like the following:
+
+```csharp
+public record Subscription
+{
+   public string Id { get; init; }
+   public NotificationFormat Format { get; init; } 
+}
+
+public record Notification
+{
+   public Subscription Subscription { get; init; }
+   public string Content { get; init; } 
+}
+```
+
+> the presented snippets use the new [C# 9 record types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record). It's not necessary; however, it's a convenient way to define immutable types.
+
+The only relevant thing in the presented snippets is probably the `NotificationFormat` enumeration type. It could have values like `EmailMessage`, `TextNotification`, `PushNotification`, `WebHook`, and many more depending on the business needs.
+
+With the above in place, procurement can define the NServiceBus message handler as follows:
+
+```csharp
+namespace Procurement.Notifications
+{
+    class PurchaseOrderCreatedHandler : IHandleMessages<IPurchaseOrderCreated>
+    {
+        readonly INotificationService notificationService;
+        public PurchaseOrderCreatedHandler(INotificationService notificationService)
+        {
+            this.notificationService = notificationService;
+        }
+
+        public async Task Handle(IPurchaseOrderCreated msg, IMessageHandlerContext ctx)
+        {
+            var subscriptionId = "procurement/purchase-order-created";
+            var subscriptions = (await notificationService.GetSubscriptionsFor(subscriptionId)).ToArray();
+            if(!subscriptions.Any())
+               return;
+
+            var formats = subscriptions.Select(s => s.Format).Distinct();
+            var formattedNotifications = new Dictionary<NotificationFormat, string>();
+            var client = new HttpClient();
+            foreach (var format in formats)
```

Comment:
The loop makes me wrinkle my nose, but this is only for 2-3 formats tops right? But wouldn't it be better to pass the list of formats to the composition service, since it's going to need to compose all the same data for each one, just in a different way, and let it return structured data that includes all the outputs in one go?

Reply:
Furthermore, why can't the Notification service, as an IT/Ops service already, do the composition? As presented, this creates coupling where the composition API service has a contract to return a string that magically fits what the notification service expects to be able to send for a given format.

Reply:
I made a note at the end that these issues will be addressed in a future post.